### ISSUE TITLE
tf.sparse.reduce_sum slower/less memory efficient than (unsorted_)segment_sum

### ISSUE THREAD

**Issue by jackd** [tensorflow/tensorflow#32763]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes (below)
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **ubuntu 16.04**
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (use command below): **v1.12.1-9365-gff401a6 1.15.0-dev20190821**
- Python version: **3.6.9**
- CUDA/cuDNN version: **10.0** / ??
- GPU model and memory: **GTX-1070**

**Current behavior**
`tf.sparse.reduce_sum` (result of a custom kernel) **50-200x slower** and 2-5x less memory efficient compared to `tf.math.unsorted_segment_sum` implementation.

**Describe the expected behavior**
Custom kernel performance should be no worse than implementations in terms of other operations.

**Code to reproduce the issue**
A very basic implementation (without support for rank > 2 or multiple summation axes) is as follows.
```python
def seg_sum(sp, axis=1, ordered=False):
    """
    math.(unsorted_)segment_sum.

    Args:
        sp: rank 2 sparse tensor
        axis: int, axis along which to sum
        ordered: if True, other axis indices are assumed to be ascending.

    Returns:
        rank 1 dense tensor equivalent to tf.sparse.reduce_sum(sp, axis=axis)
    """
    if sp.shape.ndims != 2:
        raise NotImplementedError
    other_axis = 0 if axis in (1, -1) else 1
    if ordered:
        return tf.math.segment_sum(sp.values, sp.indices[:, other_axis])
    else:
        return tf.math.unsorted_segment_sum(sp.values,
                                            sp.indices[:, other_axis],
                                            sp.dense_shape[other_axis])
```
Extending this to support full functionality should not be difficult.

Basic testing/benchmarking script:
```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np
import tensorflow as tf


def get_data(dense_shape, mean_edges=100, seed=123):
    """Generate random sparse matrix data."""
    N, M = dense_shape
    r = np.random.RandomState(seed)

    num_edges = int(mean_edges * N)

    flat_index = (r.uniform(size=num_edges) * N * M).astype(np.int64)
    # flat_index = np.concatenate([flat_index, np.arange(0, N * M, M)], axis=0)
    flat_index = np.array(sorted(set(flat_index)), dtype=np.int64)
    i, j = np.unravel_index(flat_index, (N, M))  # pylint: disable=unbalanced-tuple-unpacking
    sparse_indices = np.stack((i, j), axis=-1)
    weights = np.random.uniform(size=(i.shape[0],)).astype(np.float32)
    return sparse_indices, weights


def sparse_sum(sp, axis=1):
    """sparse.reduce_sum."""
    return tf.sparse.reduce_sum(sp, axis=axis)


def sparse_sum2(sp, axis=1):
    """sparse.reduce_sum_sparse -> to_dense."""
    return tf.sparse.to_dense(tf.sparse.reduce_sum_sparse(sp, axis=axis))


def seg_sum(sp, axis=1, ordered=False):
    """
    math.(unsorted_)segment_sum.

    Args:
        sp: rank 2 sparse tensor
        axis: int, axis along which to sum
        ordered: if True, other axis indices are assumed to be ascending.

    Returns:
        rank 1 dense tensor equivalent to tf.sparse.reduce_sum(sp, axis=axis)
    """
    if sp.shape.ndims != 2:
        raise NotImplementedError
    other_axis = 0 if axis in (1, -1) else 1
    if ordered:
        return tf.math.segment_sum(sp.values, sp.indices[:, other_axis])
    else:
        return tf.math.unsorted_segment_sum(sp.values,
                                            sp.indices[:, other_axis],
                                            sp.dense_shape[other_axis])


def compare(dense_shape, axis=1, ordered=False, **kwargs):
    sparse_indices, weights = get_data(dense_shape, **kwargs)
    sparse_indices = tf.constant(sparse_indices, dtype=tf.int64)
    weights = tf.constant(weights, dtype=tf.float32)
    sp = tf.SparseTensor(sparse_indices, weights, dense_shape)
    sparse = sparse_sum(sp, axis=axis)
    sparse2 = sparse_sum2(sp, axis=axis)
    seg = seg_sum(sp, axis=axis, ordered=ordered)
    sparse_grad, = tf.gradients(sparse, weights)
    # sparse2_grad, = tf.gradients(sparse2, weights)
    seg_grad, = tf.gradients(seg, weights)
    err = tf.reduce_max(tf.abs(seg - sparse))
    shape_err = tf.reduce_max(tf.abs(tf.shape(sparse) - tf.shape(seg)))
    err2 = tf.reduce_max(tf.abs(seg - sparse2))
    shape_err2 = tf.reduce_max(tf.abs(tf.shape(sparse2) - tf.shape(seg)))
    grad_err = tf.reduce_max(tf.abs(sparse_grad - seg_grad))
    # grad_err2 = tf.reduce_max(tf.abs(sparse2_grad - seg_grad))

    with tf.Session() as sess:
        err, shape_err, err2, shape_err2, grad_err = sess.run(
            (err, shape_err, err2, shape_err2, grad_err))
    assert (err < 1e-4)
    assert (shape_err == 0)
    assert (shape_err2 == 0)
    assert (grad_err < 1e-4)
    return err


def run_benchmarks(dense_shape, axis=1, ordered=False, **kwargs):
    sparse_indices, weights = get_data(dense_shape, **kwargs)
    sparse_indices = tf.constant(sparse_indices, dtype=tf.int64)
    weights = tf.constant(weights, dtype=tf.float32)
    sp = tf.SparseTensor(sparse_indices, weights, dense_shape)
    sparse = sparse_sum(sp, axis=axis)
    sparse_grad, = tf.gradients(sparse, weights)

    seg = seg_sum(sp, axis=axis, ordered=ordered)
    seg_grad = tf.gradients(seg, weights)

    sparse2 = sparse_sum2(sp, axis=axis)
    # sparse2_grad, = tf.gradients(sparse2, weights)
    # print(sparse2_grad)
    names = []
    time = []
    mem = []

    def update(name, result):
        time.append(result['wall_time'])
        mem.append(result['extras']['allocator_maximum_num_bytes_GPU_0_bfc'])
        names.append(name)

    with tf.Session() as sess:
        print('------------------')
        print('----- SPARSE -----')
        bm = tf.test.Benchmark()
        result = bm.run_op_benchmark(sess, (sparse, sparse_grad))
        update('sparse', result)
        # # no gradients to sparse2 - unfair comparison
        print('------------------')
        print('----- SPARSE2 -----')
        bm = tf.test.Benchmark()
        result = bm.run_op_benchmark(sess, (sparse2,))
        update('sparse2', result)
        print('------------------')
        print('----- SEG ----')
        bm = tf.test.Benchmark()
        result = bm.run_op_benchmark(sess, (seg, seg_grad))
        update('seg', result)

    time = np.array(time)
    i = np.argmin(time)
    best_time = time[i]
    print('Fastest:  {}, {}'.format(names[i], best_time))

    mem = np.array(mem)
    j = np.argmin(mem)
    best_mem = mem[j]
    print('Smallest: {}, {:.2f}mb'.format(names[j], best_mem / (1024**2)))

    print('rel time, rel mem, name')
    for name, t, m in zip(names, time, mem):
        print('{:.3f}, {:.3f}, {}'.format(t / best_time, m / best_mem, name))
    return time, mem, names


axis = 0
ordered = False
dense_shape = (int(1e4), int(1e6))
compare(axis=axis, dense_shape=dense_shape, ordered=ordered)
run_benchmarks(axis=axis, dense_shape=dense_shape, ordered=ordered)

```

**Other info / logs**
```
------------------
----- SPARSE -----
2019-09-24 16:03:05.305709: I tensorflow/core/profiler/lib/profiler_session.cc:205] Profiler session started.
entry {
  name: "TensorFlowBenchmark.run_op_benchmark"
  iters: 10
  wall_time: 0.2983849048614502
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 23998744.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_cpu"
    value {
      double_value: 4000000.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_gpu_host_bfc"
    value {
      double_value: 4.0
    }
  }
}

------------------
----- SPARSE2 -----
entry {
  name: "TensorFlowBenchmark.run_op_benchmark"
  iters: 10
  wall_time: 0.32063305377960205
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 19998956.0
    }
  }
  extras {
    key: "allocator_maximum_num_bytes_cpu"
    value {
      double_value: 11582884.0
    }
  }
}

------------------
----- SEG ----
entry {
  name: "TensorFlowBenchmark.run_op_benchmark"
  iters: 10
  wall_time: 0.0016645193099975586
  extras {
    key: "allocator_maximum_num_bytes_GPU_0_bfc"
    value {
      double_value: 7999788.0
    }
  }
}

Fastest:  seg, 0.0016645193099975586
Smallest: seg, 7.63mb
rel time, rel mem, name
179.262, 3.000, sparse
192.628, 2.500, sparse2
1.000, 1.000, seg
```

**Comment by oanush** [comment]
Please find the gist of colab  for[ TF-15.0rc1](https://colab.sandbox.google.com/gist/oanush/1d1de9cd803a5d388d4c44f2034a6697/32763.ipynb) and TF-15.0 when tried executing the code.
Thanks!

**Comment by atyshka** [comment]
Any updates on this? Testing in TF 2.3 and sparse reduce_sum is excruciatingly slow, 50ms and I only have about 50k elements. @jackd Can your workaround be extended to higher dimensions? I'm using 6D tensors

**Comment by jackd** [comment]
@atyshka what axis are you reducing over? If it's the first or last it might be easiest just to reshape to rank-2, use the above, then reshape back?

**Comment by atyshka** [comment]
Actually not one but the last two axes, which complicates things a little

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
